[ML] Deprecate stateFactory().

walterra opened the pull request:
## Summary

Follow up to #31426. This deprecates our own implementation of `stateFactory()`. This utility was able to manage shared state across angular and react code including persisting it to the URL using `AppState`. This PR replaces it with a simpler way using RxJs' Observables.

`stateFactory()` offered its own suite of methods to manage state and handle update event (get/set/watch/unwatch/changed). This is all gone now, instead we use standard RxJs's Observables. Main advantages:

- Subscribors to `BehaviorSubject` don't suffer from possible race conditions. There is now only one place where a default value gets specified and that's when the observable gets initialized. Previously we had to track up to three places to track initialization and default values to be resilient to race conditions.
- Because this is now based on observables, we can use our helper `injectObservablesAsProps`. This allowed to refactor `explorer.js` so it doesn't rely on any custom event listener code anymore.

To still be able to persist data to the url via `AppState`, a helper called `subscribeAppStateToObservable` is introduced in that PR. It's logic is much simpler than that of `stateFactory()`. It sets up a subscription to the provided observable and persists every change to the url. On initial setup it checks if there's already something persisted in the url and if so, restores that value and triggers an observable update.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- ~~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

elasticmachine commented:
Pinging @elastic/ml-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1801/)

jgowdyelastic commented on x-pack/plugins/ml/public/util/app_state_utils.js:
```diff
@@ -0,0 +1,56 @@
+    // if defaultState is an object, check if current appState has the same keys.
+    // if it's not an object, check if defaultState and current appState are of the same type.
+    if (
+      (typeof defaultState === 'object' && !_.isEqual(
```
could this `typeof` check be moved to the outer `if`, to replace the `undefined` check?

walterra replied:
I'm afraid this wouldn't work as expected: The most outer `if` checks if `defaultState` is set at all, the inner `if` contains two checks to find out if the data stored in AppState is valid:

- The one you commented on is about when `defaultState` is an object, then it gets compared to `appStates` version to see if they share the same object attributes. This is like duck typing the AppState object to find out if it has the same format then `defaultState`.
- The second check is when the state is simpler, like for `showCharts` it's just a boolean, then we compare the types of `defaultState` and `AppState`. For example, if `defaultState` for `showCharts` is `true` but for whatever reason the URL is malformed and `AppState` returns `123` it would fall back to `defaultState`.

By allowing these simpler values in AppState (`stateFactory()` always required an object) our URLs got a bit shorter, for example the value stored in the URL for `showCharts` now looks like this: `mlCheckboxShowCharts:!t`, previously it was something like `mlCheckboxShowCharts:(showCharts:!t)`

jgowdyelastic commented on x-pack/plugins/ml/public/util/app_state_utils.js:
```diff
@@ -0,0 +1,56 @@
+  return appState;
+}
+
+export function subscribeAppStateToObservable(AppState, appStateName, o$, $rootScope) {
```
rather than passing in `$rootScope` could this take a `refresh` function (or some better name) instead which is passed `$rootScope.$applyAsync`
That way it's not tied to angular.
It would probably need to be wrapped in a `typeof === 'function'` check for safety

walterra replied:
Good idea, I changed that last argument to just `callback` which gets called on each subscription trigger in 4a3578b.

jgowdyelastic commented on x-pack/plugins/ml/public/util/__tests__/app_state_utils.js:
```diff
@@ -0,0 +1,64 @@
+describe('ML - initializeAppState', () => {
+  let AppState;
+
+  beforeEach(ngMock.module('kibana', function (stateManagementConfigProvider) {
```
could this be an arrow function?

walterra replied:
Done in 4a3578b.

jgowdyelastic commented on x-pack/plugins/ml/public/util/__tests__/app_state_utils.js:
```diff
@@ -0,0 +1,64 @@
+  let AppState;
+  let $rootScope;
+
+  beforeEach(ngMock.module('kibana', function (stateManagementConfigProvider) {
```
could this be an arrow function?

walterra replied:
Done in 4a3578b.

walterra commented:
@alvarezmelissa87 @jgowdyelastic 
- Fixed restoring a swimlane selection from URL in fbef324. 
- Fixed restoring the view by selection from URL in ad34d28.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1995/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2003/)
